src/device/router_device_routing.erl

```diff
@@ -944,6 +967,61 @@ find_device(PubKeyBin, DevAddr, MIC, Payload, Chain) ->
             {Device, NwkSKey}
     end.
 
+%%%-------------------------------------------------------------------
+%% @doc
+%% If there are devaddr collisions, we don't know the correct device until we
+%% get the whole packet with the payload. Once we have that, the device can tell
+%% routing the a packet hash belongs to it. That reservation will only last a
+%% short while. But it might be enough to have subsequent offers lookup the
+%% correct device by the time they come through.
+%% @end
+%%%-------------------------------------------------------------------
+-spec cache_device_for_hash(PHash :: binary(), Device :: router_device:device()) -> ok.
+cache_device_for_hash(PHash, Device) ->
+    ok = e2qc_nif:put(
```

Comment:
Using `nif` so you don't have to use a function I am guessing?

Reply:
yeah, for this putting we don't get the same benefit as with getting. 
It felt more symmetrical this way.